### ISSUE TITLE
"Null Output" showing as name of audio device in Sound

### ISSUE THREAD

**Issue by xmbwd** [muammar/mkchromecast#62]
When I switch from one Chromecast device to the other, a new sink is created, sometimes called mkchromecast and sometimes called Null Output. Sometimes, they are both called Null Output. This makes it difficult to determine where the sound will be routed. Any way to remedy?

![screenshot from 2017-04-22 11-06-17](https://cloud.githubusercontent.com/assets/12945044/25309840/57c026c0-278c-11e7-9565-78c90556dedd.png)

**Comment by muammar** [comment]
I may have fixed it here https://github.com/muammar/mkchromecast/commit/da8d1f49f1319c3ca1c7395e1b97f6fe27abfe4b. But as I was working in the `sonos` branch, it is not still there in `devel` or `master`. Sorry about that!. I will try to merge `sonos` back to `devel` maybe later this week.

**Comment by xmbwd** [comment]
Wanted to check in on this -- is in `dev` or `master`?  After a few days of play, I get six different `null outputs` in my sound prefs, making it hard to figure out which one is live.  Appreciate it.

**Comment by muammar** [comment]
It is in the `sonos` branch if you want to test the fix. Thank you!.

**Comment by xmbwd** [comment]
It looks like the `sonos` branch is only available on `OSX`.  I'm running `Linux`:

```
sudo python3 setup.py install
Traceback (most recent call last):
  File "setup.py", line 18, in <module>
    from setuptools import setup
ImportError: No module named 'setuptools'
mbwd@mbwd-iMac:~/Downloads/mkchromecast-sonos$ sudo apt install setuptools
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package setuptools
mbwd@mbwd-iMac:~/Downloads/mkchromecast-sonos$ sudo apt-get install python-pip
Reading package lists... Done
Building dependency tree       
Reading state information... Done
python-pip is already the newest version (8.1.1-2ubuntu0.4).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
mbwd@mbwd-iMac:~/Downloads/mkchromecast-sonos$ sudo apt-get install python3-setuptools
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Suggested packages:
  python-setuptools-doc
The following NEW packages will be installed:
  python3-setuptools
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 88.0 kB of archives.
After this operation, 439 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu xenial/main amd64 python3-setuptools all 20.7.0-1 [88.0 kB]
Fetched 88.0 kB in 1s (84.2 kB/s)                    
Selecting previously unselected package python3-setuptools.
(Reading database ... 356009 files and directories currently installed.)
Preparing to unpack .../python3-setuptools_20.7.0-1_all.deb ...
Unpacking python3-setuptools (20.7.0-1) ...
Setting up python3-setuptools (20.7.0-1) ...
mbwd@mbwd-iMac:~/Downloads/mkchromecast-sonos$ sudo python3 setup.py install
warning: no directories found matching 'doc/_static'
warning: no directories found matching 'doc/_templates'
warning: no previously-included files matching '.DS_Store' found anywhere in distribution
warning: no previously-included files matching '*.pyc' found anywhere in distribution
warning: no previously-included files matching '*.pyo' found anywhere in distribution
warning: no previously-included files matching '*.so' found anywhere in distribution
warning: no previously-included files matching '*.dSYM' found anywhere in distribution
warning: no previously-included files matching '*.dSYM/*' found anywhere in distribution
warning: no previously-included files matching '*__pycache__*' found anywhere in distribution
no previously-included directories found matching 'examples/*/build'
no previously-included directories found matching 'examples/*/dist'
no previously-included directories found matching 'examples/*/*/build'
no previously-included directories found matching 'examples/*/*/dist'
no previously-included directories found matching 'examples/*/*/*/build'
no previously-included directories found matching 'examples/*/*/*/dist'
Traceback (most recent call last):
  File "/usr/lib/python3.5/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.5/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/lib/python3/dist-packages/setuptools/command/bdist_egg.py", line 161, in run
    cmd = self.call_command('install_lib', warn_dir=0)
  File "/usr/lib/python3/dist-packages/setuptools/command/bdist_egg.py", line 147, in call_command
    self.run_command(cmdname)
  File "/usr/lib/python3.5/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/tmp/easy_install-zl3mnmi1/py2app-0.12/setup.py", line 686, in run
distutils.errors.DistutilsPlatformError: This distribution is only supported on MacOSX

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 154, in save_modules
    yield saved
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 243, in run_setup
    DirectorySandbox(setup_dir).run(runner)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 273, in run
    return func()
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 242, in runner
    _execfile(setup_script, ns)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 46, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-zl3mnmi1/py2app-0.12/setup.py", line 789, in <module>
  File "/usr/lib/python3.5/distutils/core.py", line 163, in setup
    raise SystemExit("error: " + str(msg))
SystemExit: error: This distribution is only supported on MacOSX

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 1087, in run_setup
    run_setup(setup_script, args)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 246, in run_setup
    raise
  File "/usr/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/usr/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 166, in save_modules
    saved_exc.resume()
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 141, in resume
    six.reraise(type, exc, self._tb)
  File "/usr/lib/python3/dist-packages/pkg_resources/_vendor/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 154, in save_modules
    yield saved
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 243, in run_setup
    DirectorySandbox(setup_dir).run(runner)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 273, in run
    return func()
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 242, in runner
    _execfile(setup_script, ns)
  File "/usr/lib/python3/dist-packages/setuptools/sandbox.py", line 46, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-zl3mnmi1/py2app-0.12/setup.py", line 789, in <module>
  File "/usr/lib/python3.5/distutils/core.py", line 163, in setup
    raise SystemExit("error: " + str(msg))
SystemExit: error: This distribution is only supported on MacOSX

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 73, in <module>
    setup_requires=['py2app'],
  File "/usr/lib/python3.5/distutils/core.py", line 108, in setup
    _setup_distribution = dist = klass(attrs)
  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 269, in __init__
    self.fetch_build_eggs(attrs['setup_requires'])
  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 313, in fetch_build_eggs
    replace_conflicting=True,
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 826, in resolve
    dist = best[req.key] = env.best_match(req, ws, installer)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1092, in best_match
    return self.obtain(req, installer)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1104, in obtain
    return installer(requirement)
  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 380, in fetch_build_egg
    return cmd.easy_install(req)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 663, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 693, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 873, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 1101, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/usr/lib/python3/dist-packages/setuptools/command/easy_install.py", line 1089, in run_setup
    raise DistutilsError("Setup script exited with %s" % (v.args[0],))
distutils.errors.DistutilsError: Setup script exited with error: This distribution is only supported on MacOSX
```

**Comment by muammar** [comment]
The `setup.py` is only there for macOs, that's true. But you can just execute `python mkchromecast.py` from the cloned repository. I should fix the `setup.py` file.

**Comment by xmbwd** [comment]
Oops.  Got it!  Was using `setup.py`.

**Comment by muammar** [comment]
> Oops. Got it! Was using setup.py.

There are many things I still have to fix :) sorry about that!. Hope that this change fixes the issue.

**Comment by xmbwd** [comment]
Unfortunately, when running `sonos` using `python mkchromecast.py`, it does not appear to fix the `null output` issue.  Running it in `terminal` does not start the applet, should it?

**Comment by muammar** [comment]
If you are interested in the systray, pass the `-t` option: `python mkchromecast.py -t`.

**Comment by muammar** [comment]
The change introduced will not create another sink if there is already one named `mkchromecast`. If this is not working, then something a bit more aggressive should be done (`pulseaudio -k` [see this](https://askubuntu.com/questions/650278/automate-creation-of-pulseaudio-null-sinks)). But not sure if that is going to be very welcomed.

**Comment by xmbwd** [comment]
Let me restart and see what happens.  There is no sink named `mkchromecast`, but that may be residual nulls.

**Comment by muammar** [comment]
> Let me restart and see what happens. There is no sink named mkchromecast, but that may be residual nulls.

Yes, that may help to sort this out. On my case, if I kill the application suddenly (using `kill -9 processnumber`), no new sinks are created although sometimes the only one working after I restart mkchromecast takes the` Null Output` name. But that is better than having many of them.

**Comment by xmbwd** [comment]
I restarted and switched sinks.  Looks like it is working.  I'm not sure why, but sometimes it shows up as `mkchromecast` and sometimes `null output` -- sometimes simultaneously (see picture below).  In `Sound` and `pavucontrol` it seems to usually have `mkchromecast`.  Using `sound switcher indicator` it seems to usually show `null output`, regardless of the others. 

![screenshot from 2017-04-30 12-16-40](https://cloud.githubusercontent.com/assets/12945044/25567289/3a6b603a-2d9f-11e7-94d8-9da18f1b7d4f.png)

**Comment by muammar** [comment]
Thank you. It is exactly the same that happens for me. I will keep using it to see if that can also be improved. Thanks for reporting back!.

**Comment by muammar** [comment]
Are you still having the issue of multiple sinks or only one sink but having sometimes the `Null` name?.

**Comment by xmbwd** [comment]
Sorry man.  I couldn't get the delay down short enough so I stopped using it.  If I recall correctly, you had fixed that issue in one of the builds.  

I should note that I tried to install `mkchromecast` on a few temporary Linux USBs and was unable to.  There is no install on the main repositories and even the deb didn't work for want of a number of dependencies (not the list of ones you have on the main site, additional ones).
